core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -1076,15 +1076,39 @@ public static String substring(String c, int s, int l) {
   }
 
   public static String substring(String c, int s, long l) {
-    return substring(c, s, clamp(l));
+    if (s < 0) {
+      if (l > 0 && l + s > 0) {
+        l += s;
+        return substring(c, 0, l);
+      }
+    }
+    int l0 = clamp(l);
+    return SqlFunctions.substring(c, s, l0);
   }
 
   public static String substring(String c, long s, int l) {
-    return substring(c, clamp(s), l);
+    if (s < 0) {
+      long s0 = s + l;
```

Comment:
it makes it easier to review if you are consistent in your implementations
elsewhere you used l += s

Reply:
done